### PR TITLE
smf: fix null pointer for unitized state machine

### PR DESCRIPTION
**Pull request by Janbk**
Before calling smf_set_initial, setting a state by smf_set_state crashes with a null pointer exception.

### CODE REVIEW

File: lib/smf/smf.c [hunk 1]
```diff
@@ -193,8 +193,7 @@ static bool smf_execute_all_exit_actions(struct smf_ctx *const ctx, const struct
 	struct internal_ctx *const internal = (void *)&ctx->internal;
 
 	for (const struct smf_state *to_execute = ctx->current;
-	     to_execute != NULL && to_execute != topmost;
-	     to_execute = to_execute->parent) {
+	     to_execute != NULL && to_execute != topmost; to_execute = to_execute->parent) {
```
Comment (kartben) [inline lib/smf/smf.c:196]: this formatting change seems unnecessary?

**Comment by maass-hamburg** [comment]
@Janbk `smf_set_initial` initials the state mashine, therefore it should always be called before `smf_set_state`. You could add a check to `smf_set_state` (and `smf_run_state`) to check if  `ctx->current` had been set.
Also `smf_set_state` should only be called inside a state.

**Comment by maass-hamburg** [comment]
You might want to take a look of the other parts in the zephyr kernel or in the docs to understand how to use the smf.

**Comment by glenn-andrews** [comment]
> @Janbk `smf_set_initial` initials the state mashine, therefore it should always be called before `smf_set_state`. You could add a check to `smf_set_state` (and `smf_run_state`) to check if `ctx->current` had been set. Also `smf_set_state` should only be called inside a state.

That check would impart a small but nonzero runtime penalty which I'd prefer to avoid. If there's any chance of null pointer de-reference occurring during normal operation, we should fix it, but I think you're safe unless you deliberately set the state to NULL.

**Review by glenn-andrews — CHANGES_REQUESTED** [review]
This seems unnecessary as the state machine must _never_ be used without `smf_set_initial` being called first. It won't just be the `current` pointer that is incorrect.

**Comment by Janbk** [comment]
@glenn-andrews @maass-hamburg @kartben Thank you all for the supporting, really appreciate your feedback.

As the name of the PR suggests,  it is a fix for a wrong used state machine function. It is more or less the same reasoning as the handling for "Calling %s from exit action". I rewrote this PR and introduced a new internal state, defining the run phase. I hope it is Ok to more or less restart the PR here with the new solution.

At first I tried to prevent introducing new internal states, because in the long run it increases complexity. On the other side, this flag is well defined. In addition we only had 4 internal bits, so I assume the code will not use a bigger memory for the internal_ctx on all hardware architectures.

What do you think about this? If the answer is, the user is responsible for reading the documentation, I close the PR with no problem.

File: lib/smf/smf.c (hunk 1 above)
Comment (Janbk) [reply lib/smf/smf.c]: Removed, sorry for the unintended change.

**Comment by glenn-andrews** [comment]
> @glenn-andrews @maass-hamburg @kartben Thank you all for the supporting, really appreciate your feedback.
> 
> As the name of the PR suggests, it is a fix for a wrong used state machine function. It is more or less the same reasoning as the handling for "Calling %s from exit action". I rewrote this PR and introduced a new internal state, defining the run phase. I hope it is Ok to more or less restart the PR here with the new solution.
> 
> At first I tried to prevent introducing new internal states, because in the long run it increases complexity. On the other side, this flag is well defined. In addition we only had 4 internal bits, so I assume the code will not use a bigger memory for the internal_ctx on all hardware architectures.
> 
> What do you think about this? If the answer is, the user is responsible for reading the documentation, I close the PR with no problem.

Good point about the exit action log message. 

I think we put that in because from a state machine standpoint you CAN put a choice of destination in an exit action (using a junction or choice pseudostate), but SMF does not support it - you have to decide your destination within the state handler and call `smf_set_state` with the appropriate destination from there.

Part of me thinks more protection is good, part of me thinks more protection slows things down when used correctly. I'd love to hear other perspectives on this.

**Comment by glenn-andrews** [comment]
I'm thinking unless there's a good reason to add the error, to close this PR.

**Comment by Janbk** [comment]
Closed the PR because of your comments and because it would break the init of USB Power Delivery Protocol Layer. 

Call stack would be: file `zephyr/subsys/usb/usb_c/usbc_prl.c`
[prl_run](https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/usb/usb_c/usbc_prl.c#L292)
-> calls [prl_init](https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/usb/usb_c/usbc_prl.c#L305)
-> calls [prl_hr_set_state](https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/usb/usb_c/usbc_prl.c#L550)
-> calls [smf_set_state](https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/usb/usb_c/usbc_prl.c#L415).

**Comment by glenn-andrews** [comment]
> Closed the PR because of your comments and because it would break the init of USB Power Delivery Protocol Layer.
> 
> Call stack would be: file `zephyr/subsys/usb/usb_c/usbc_prl.c` [prl_run](https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/usb/usb_c/usbc_prl.c#L292) -> calls [prl_init](https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/usb/usb_c/usbc_prl.c#L305) -> calls [prl_hr_set_state](https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/usb/usb_c/usbc_prl.c#L550) -> calls [smf_set_state](https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/usb/usb_c/usbc_prl.c#L415).

`smf_set_initial()` is called by `prl_subsys_init()` which is called by `usbc_subsys_init()` which is the init function for the USB-C subsystem called when all drivers are initialized, so I think we're safe.

If there's a path available where it's NOT being called, then we have a bug that needs to be flagged.

**Comment by glenn-andrews** [comment]
@Janbk I want to thank you for adding this PR and being so receptive to (hopefully) constructive criticism. 

Welcome to Zephyr and I hope you provide many more PRs for the project!

**Pull request closed without merging**